### PR TITLE
[CoSim] strong coupling MPI

### PR DESCRIPTION
**Pull request by philbucher**
With these changes the CoSimApp is 100% MPI-compatible :v:

This basically requires wrappers around the Convergence Accelerators and the Convergence Criteria. This has the advantage that those classes can then also be directly used in python, which is a requirement from Gent

Outlook: With #5489 we will finally be able to use a subset of cores for a solver in co-sim, i.e. running Fluid in 100 cores and structure in 20 (launching Kratos with 100 cores)

Contains #5492

part of #5140 

TODO for this to be ready:
- [x] testing on the cluster
- [x] finish writing tests
- [x] fix minor TODOs in code
- [x] #6001 should be merged first

### CODE REVIEW

**Comment by AndreasWinterstein** [comment]
Tested on the cluster. Works.

**Comment by philbucher** [comment]
this is ready now
I added tests for checking the implementation with and without MPI

**Comment by adityaghantasala** [comment]
I will have a look over the weekend .. !

File: applications/CoSimulationApplication/python_scripts/convergence_accelerators/convergence_accelerator_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+        if self.gather_scatter_required:
+            residual = np.array(np.concatenate(self.data_comm.GathervDoubles(residual, 0)))
+            input_data_for_acc = np.array(np.concatenate(self.data_comm.GathervDoubles(input_data_for_acc, 0)))
+
+        if self.executing_rank:
+            updated_data = input_data_for_acc + self.conv_acc.UpdateSolution(residual, input_data_for_acc)
+
+        if self.gather_scatter_required:
+            if self.my_pid == 0:
```
Comment (adityaghantasala) [inline applications/CoSimulationApplication/python_scripts/convergence_accelerators/convergence_accelerator_wrapper.py]:
```suggestion
            if self.executing_rank:
```

File: applications/CoSimulationApplication/python_scripts/convergence_criteria/convergence_criteria_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    def IsConverged(self):
+        current_data = self.interface_data.GetData()
+        residual = current_data - self.input_data
+
+        if self.interface_data.IsDistributed():
+            residual = np.array(np.concatenate(self.data_comm.GathervDoubles(residual, 0)))
+            current_data = np.array(np.concatenate(self.data_comm.GathervDoubles(current_data, 0)))
+
+        is_converged = 0
+        if self.my_pid == 0:
```
Comment (adityaghantasala) [inline applications/CoSimulationApplication/python_scripts/convergence_criteria/convergence_criteria_wrapper.py]: can you also use the same variable as in the accelerator 

`self.executing_rank = conv_acc_supports_dist_data or (self.my_pid == 0)`

Comment (philbucher) [reply applications/CoSimulationApplication/python_scripts/convergence_criteria/convergence_criteria_wrapper.py]: done

File: applications/CoSimulationApplication/python_scripts/convergence_accelerators/convergence_accelerator_wrapper.py (hunk 1 above)
Comment (philbucher) [reply applications/CoSimulationApplication/python_scripts/convergence_accelerators/convergence_accelerator_wrapper.py]: done

**Comment by philbucher** [comment]
suggestions applied @adityaghantasala

**Review by adityaghantasala — APPROVED** [review]

**Comment by adityaghantasala** [comment]
Please wait till evening to merge !

**Pull request merged**